src-input/duk_bi_encoding.c

```diff
@@ -0,0 +1,333 @@
+#define DUK__U8_RETRY      3  /* retry byte, constraint: RETRY > ERROR */
+
+DUK_LOCAL void duk__utf8_decode_init(duk__decode_context *dec_ctx) {
+	dec_ctx->codepoint = 0x0000;
+	dec_ctx->needed = 0;
+	dec_ctx->seen = 0;
+	dec_ctx->upper = 0xbf;
+	dec_ctx->lower = 0x80;
+	dec_ctx->bom_seen = 0;
+}
+
+DUK_LOCAL int duk__utf8_decode_next(duk__decode_context *dec_ctx, duk_uint8_t byte, duk_codepoint_t *cp) {
+	/*
+	 *  UTF-8 algorithm based on the Encoding specification:
+	 *  https://encoding.spec.whatwg.org/#utf-8-decoder
+	 */
+
+	if (dec_ctx->needed == 0) {
+		/* process initial byte */
+		if (byte <= 0x7f) {
+			/* U+0000 - U+007F, 1 byte (ASCII) */
+			*cp = (duk_codepoint_t) byte;
+			return DUK__U8_CODEPOINT;
+		} else if (byte >= 0xc2 && byte <= 0xdf) {
+			/* U+0080 - U+07FF, 2 bytes */
+			dec_ctx->needed = 2;
+			dec_ctx->codepoint = byte & 0x1f;
+			return DUK__U8_CONTINUE;
+		} else if (byte >= 0xe0 && byte <= 0xef) {
+			/* U+0800 - U+FFFF, 3 bytes */
+			if (byte == 0xe0) {
+				dec_ctx->lower = 0xa0;
+			} else if (byte == 0xed) {
+				dec_ctx->upper = 0x9f;
+			}
+			dec_ctx->needed = 2;
+			dec_ctx->codepoint = byte & 0xf;
+			return DUK__U8_CONTINUE;
+		} else if (byte >= 0xf0 && byte <= 0xf4) {
+			/* U+010000 - U+10FFFF, 4 bytes */
+			if (byte == 0xf0) {
+				dec_ctx->lower = 0x90;
+			} else if (byte == 0xf4) {
+				dec_ctx->upper = 0x8f;
+			}
+			dec_ctx->needed = 3;
+			dec_ctx->codepoint = byte & 0x7;
+			return DUK__U8_CONTINUE;
+		} else {
+			/* not a valid UTF-8 initial byte */
+			return DUK__U8_ERROR;
+		}
+	} else {
+		/* process continuation byte */
+		if (byte >= dec_ctx->lower && byte <= dec_ctx->upper) {
+			dec_ctx->lower = 0x80;
+			dec_ctx->upper = 0xbf;
+			dec_ctx->codepoint = (dec_ctx->codepoint << 6) | (byte & 0x3f);
+			if (++dec_ctx->seen < dec_ctx->needed) {
+				/* need more bytes */
+				return DUK__U8_CONTINUE;
+			} else if (dec_ctx->codepoint < 0xd800 || dec_ctx->codepoint > 0xdfff) {
+				/* got a codepoint */
+				*cp = dec_ctx->codepoint;
+				dec_ctx->codepoint = 0x0000;
+				dec_ctx->needed = 0;
+				dec_ctx->seen = 0;
+				return DUK__U8_CODEPOINT;
+			} else {
+				/* surrogate, not valid in UTF-8 */
+				dec_ctx->codepoint = 0x0000;
+				dec_ctx->needed = 0;
+				dec_ctx->seen = 0;
+				return DUK__U8_ERROR;
+			}
+		} else {
+			/* We've just seen an invalid UTF-8 continuation byte.  This might
+			 * be the initial byte of the next character; if we return a plain
+			 * error status and the decoder is in replacement mode, the character
+			 * will be masked.  We still need to alert the caller to the error
+			 * though.
+			 */
+			dec_ctx->codepoint = 0x0000;
+			dec_ctx->needed = 0;
+			dec_ctx->seen = 0;
+			dec_ctx->lower = 0x80;
+			dec_ctx->upper = 0xbf;
+			return DUK__U8_RETRY;
+		}
+	}
+}
+
+DUK_LOCAL void duk__utf8_encode_char(void *udata, duk_codepoint_t codepoint) {
+	duk__encode_context *enc_ctx;
+
+	enc_ctx = (duk__encode_context *) udata;
+	if (codepoint > 0x10ffff) {
```

Comment:
Here's an example that I encountered a while back which still gives me the creeps:

```c
#include <stdio.h>
#include <stdint.h>
int main(int argc, char *argv[]) {
	(void) argc; (void) argv;
	int64_t test = -0x80000000;
	printf("%lld\n", (long long) test);
	return 0;
}
```

This prints:

```
2147483648
```

A positive rather than negative value. The situation remains the same using the constant `-0x80000000L` but is fixed by using `-0x80000000LL`.

The reason? As far as I can tell, the literal is parsed as an unary minus followed by a non-negative constant. `0x80000000` parses as an unsigned integer because it doesn't fit into a signed integer (but fits into 32 bits), so unary minus of that is still a positive (32-bit unsigned) value. *That* then gets coerced into signed 64 bits.

This was very much fun figuring out ;)

Anyway, here too the issue is avoided if the constant matches the width *and* sign of the intended result instead of relying on automatic conversion.